shenyu-web/src/main/java/org/apache/shenyu/web/handler/ShenyuWebHandler.java

```diff
@@ -94,12 +104,41 @@ public void putExtPlugins(final List<ShenyuPlugin> extPlugins) {
                 .filter(e -> plugins.stream().noneMatch(plugin -> plugin.named().equals(e.named())))
                 .collect(Collectors.toList());
         if (CollectionUtils.isNotEmpty(shenyuPlugins)) {
-            shenyuPlugins.forEach(plugin -> {
-                plugins.add(plugin);
-                LOG.info("shenyu auto add extends plugins:{}", plugin.named());
-            });
+            shenyuPlugins.forEach(plugin -> LOG.info("shenyu auto add extends plugins:{}", plugin.named()));
+            shenyuPlugins.addAll(plugins);
+            this.plugins = sortPlugins(shenyuPlugins);
         }
     }
+    
+    /**
+     * listen sort plugin event and sort plugin.
+     *
+     * @param event sort plugin event
+     */
+    @Override
+    public void onApplicationEvent(final SortPluginEvent event) {
+        // copy a new one, or there will be concurrency problems
+        this.plugins = sortPlugins(new ArrayList<>(plugins));
+    }
+
+    /**
+     * sort plugins.
+     *
+     * @param list list of plugin
+     * @return sorted list
+     */
+    private List<ShenyuPlugin> sortPlugins(final List<ShenyuPlugin> list) {
+        Map<String, Integer> pluginSortMap = list.stream().collect(Collectors.toMap(ShenyuPlugin::named, plugin -> {
+            PluginData pluginData = BaseDataCache.getInstance().obtainPluginData(plugin.named());
+            if (pluginData == null) {
```

Comment:
used Objects

Reply:
GOOD IDEA.  I replace it to Optional.ofNullable(pluginData).map(PluginData::getSort).orElse(plugin.getOrder());